### ISSUE TITLE
Use of `this.reference` in `Array.prototype.map` refers to implicit global, not outer scope

### ISSUE THREAD

**Issue by moll** [Hardmath123/nearley#123]
Hey,

Noticed https://github.com/Hardmath123/nearley/blob/master/lib/nearley.js#L167 doesn't pass the outer scope's `this` to `Array.prototype.map`, therefore failing in strict mode where `this` is not the global object, but `null`.

**Comment by Hardmath123** [comment]
Sorry, I'm not sure I understand what you're saying. Could you please clarify what you mean?

**Comment by moll** [comment]
The line I referred to above uses `this`. This refers to the global object (`window` on Browsers and `global` on Node.js) because the `Array.prototype.map` doesn't call it in any particular context. (Context means the `this` keyword). Pass `this` as the 3rd arg to `Array.prototype.map` and you can remove the unused `var me = this` line above.

**Comment by moll** [comment]
I mean the way the code calls `Array.prototype.map` doesn't call it in any context. If you pass the 3rd arg, only then will it be in the context I think you meant it to be in.

**Comment by moll** [comment]
Hehe, sorry, not the 3rd argument, but the 2nd argument. ^_^

```js
[1, 2, 3].map(mapFn, thisArg)
```

**Comment by moll** [comment]
Run your tests with --use-strict if you use Node.js. That will catch bugs like these. ;)

**Comment by bd82** [comment]
http://stackoverflow.com/questions/80084/in-javascript-why-is-the-this-operator-inconsistent

**Comment by Hardmath123** [comment]
Hmm, I *still* don't get it. I don't see a `map` on the line you pointed out; I assume you mean the `forEach` on line 145? In that case we're in a `State.prototype` method, so since it's called as an object method, shouldn't `this` be bound to the `State` in question (the receiving object)?

Maybe providing some sample input to `nearley` that causes a failure would help me understand what's up. Sorry if I'm being dense.

**Comment by moll** [comment]
Haha, no prob, as it turns out I too am slightly off — every time I wrote `Array.prototype.map`, I was really thinking of `Array.prototype.forEach`. :D

That's the thing — you pass an anonymous function to `forEach` which is no longer in `State.prototype`'s context. Pass `this` as the 2nd arg to [`forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) to be in there.

As to make it crash rather than fail silently, run any code that uses Nearley in strict mode:
```sh
node --use-strict parser.js
```

It should crash with `TypeError: Cannot read property 'reference' of undefined`.

Or just print out `this` at the line I referenced and you'll see it refers to `global`, not an instance of `State.prototype`.
